contracts/modules/v2/StakingPool.sol

```diff
@@ -349,71 +378,122 @@ contract StakingPool is ERC20 {
 
   /* callable by pool owner */
 
-  function addProduct() external {
+  function addProduct() external onlyManager {
 
     //
 
   }
 
-  function removeProduct() external {
+  function removeProduct() external onlyManager {
 
     //
 
   }
 
-  function setWeights() external {
+  function setWeights() external onlyManager {
 
     //
 
   }
 
+  function setTargetPrice(uint productId, uint targetPrice) external onlyManager {
+    require(targetPrice >= GLOBAL_MIN_PRICE, "StakingPool: Target price must be greater than global min price");
+    targetPrices[productId] = targetPrice;
+  }
+
   /* VIEWS */
-  
+
   /* ========== PRICE CALCULATION ========== */
 
-  function getUsedCapacity(uint productId) public view returns (uint) {
-    return 0;
+  function calculatePremium(uint pricePercentage, uint coverAmount, uint period) public pure returns (uint) {
+    return pricePercentage * coverAmount / MAX_PRICE_PERCENTAGE * period / 365 days;
   }
 
-  function getCapacity(uint productId, uint capacityFactor) public view returns (uint) {
-    return 0;
-  }
+  uint public constant SURGE_THRESHOLD = 8e17;
+  uint public constant BASE_SURGE_LOADING = 1e16;
 
-  function calculatePremium(uint pricePercentage, uint coverAmount, uint period) public pure returns (uint) {
-    return pricePercentage * coverAmount / MAX_PRICE_PERCENTAGE * period / 365 days;
+
+  function getActualPriceAndUpdateBasePrice(
+    uint productId,
+    uint amount,
+    uint activeCover,
+    uint capacity,
+    uint initialPrice
+  ) internal returns (uint) {
+
+
+    (uint actualPrice, uint basePrice) = getPrices(productId, amount, activeCover, capacity, initialPrice);
+    // store the last base price
+    lastPrices[productId] = LastPrice(
+      uint96(basePrice),
+      uint32(block.timestamp
+    ));
+
+    return actualPrice;
   }
 
-  function calculatePrice(
+  function getPrices(
+    uint productId,
     uint amount,
-    uint basePrice,
     uint activeCover,
-    uint capacity
-  ) public pure returns (uint) {
+    uint capacity,
+    uint initialPrice
+  ) public view returns  (uint actualPrice, uint basePrice) {
+    uint96 lastPrice = lastPrices[productId].value;
+    basePrice = interpolatePrice(
+      lastPrice != 0 ? lastPrice : initialPrice,
+      targetPrices[productId],
+      lastPrices[productId].lastUpdateTime,
```

Comment:
This reads from storage twice. Store the whole LastPrice struct in memory instead of just the value and use the lastPrice variable from memory to get both value and lastUpdateTime.